### PR TITLE
[#10920] Add front-end tests for Session Edit Page

### PR DESCRIPTION
**Pull request by ypinhsuan**
Part of #10920

Outline of Solution
- Add snapshot tests with default and data loaded
- Add API tests for getting course, feedback session, students and instructors
- Add tests to check behaviour of button

### CODE REVIEW

**Comment by rrtheonlyone** [comment]
@ypinhsuan nice work! Please fix the lint checks before we proceed to review

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts [hunk 1]
```diff
@@ -1,33 +1,406 @@
...
+    expect(fixture).toMatchSnapshot();
+  });
+
+  it('should snap with feedback session questions', () => {
+    component.questionEditFormModels = [testQuestionEditFormModel1, testQuestionEditFormModel2];
+    component.isLoadingFeedbackSession = false;
+    component.isLoadingFeedbackQuestions = false;
+    fixture.detectChanges();
+    expect(fixture).toMatchSnapshot();
+  });
```
Comment (halfwhole) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:247]: Could we also set the component's course ID/course name/feedback session name, to verify that it shows up as intended?

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts [hunk 2]
```diff
@@ -1,33 +1,406 @@
...
+
+  it('should snap when feedback session failed to load', () => {
+    component.hasLoadingFeedbackSessionFailed = true;
+    fixture.detectChanges();
+    expect(fixture).toMatchSnapshot();
+  });
+
```
Comment (halfwhole) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:262]: Snapshots with other boolean variables would be welcome too, e.g. is copying question, or whether loading the feedback questions has failed

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts [hunk 3]
```diff
@@ -1,33 +1,406 @@
...
+  it('should expand all questions when at least one question is collapsed', () => {
+    testQuestionEditFormModel1.isCollapsed = true;
+    testQuestionEditFormModel2.isCollapsed = false;
+    component.questionEditFormModels = [testQuestionEditFormModel1, testQuestionEditFormModel2];
+    component.isLoadingFeedbackQuestions = false;
+    fixture.detectChanges();
+
+    const button: any = fixture.debugElement.nativeElement.querySelector('#btn-collapse-expand');
+    button.click();
```
Comment (halfwhole) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:358]: Would have thought to test `collapseAll()` and `expandAll()` individually, but this also looks good to me!

**Review by halfwhole — CHANGES_REQUESTED** [review]
This already looks solid, though we should still be able to add further things for more completeness:
- It seems good to test `copyCurrentSession()` as well, though that might be more tricky because we have to mock the `modalRef`. Would it be feasible?
- It also would be good to have a few more types of feedback questions beyond `FeedbackQuestionType.TEXT`

**Comment by ypinhsuan** [comment]
I have added snapshots with other boolean variables and a test for `copyCurrentSession()`, but I am not sure whether the mock `modalRef` is used correctly

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts [hunk 4]
```diff
@@ -1,33 +1,490 @@
...
   let component: InstructorSessionEditPageComponent;
   let fixture: ComponentFixture<InstructorSessionEditPageComponent>;
+  let feedbackSessionsService: FeedbackSessionsService;
+  let feedbackQuestionsService: FeedbackQuestionsService;
+  let courseService: CourseService;
+  let studentService: StudentService;
+  let instructorService: InstructorService;
+  let navigationService: NavigationService;
+  let ngbModal: NgbModal;
+  const mockModalRef: MockNgbModalRef = new MockNgbModalRef();
```
Comment (halfwhole) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:249]: We can shift this to within the test case itself, seems it looks like it's used only there

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts [hunk 5]
```diff
@@ -1,33 +1,490 @@
...
 import { InstructorSessionEditPageModule } from './instructor-session-edit-page.module';
+import Spy = jasmine.Spy;
+
+class MockNgbModalRef {
+  componentInstance: any = {
+    newFeedbackSessionName: '',
+    courseCandidates: [],
+    sessionToCopyCourseId: '',
+  };
+  result: Promise<any> = Promise.resolve();
```
Comment (halfwhole) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:51]: Can probably shift this class to within the test case too?
Also we might want to test `copyCurrentSession()` further by spying on say `createFeedbackSession()` for the feedback sessions service. That could be quite involved though, would have to change a couple things and return a promise for `copyCurrentSession()` so we know when it's done. Then this modalRef's `result` would need to give the appropriate desired result also

**Review by halfwhole — COMMENTED** [review]
Usage of mock `modalRef` looks good

**Review by halfwhole — COMMENTED** [review]
LGTM, though when I run it locally there are these console errors/warnings about `Error during cleanup of component` and `TypeError: observer.disconnect is not a function`? They don't seem to affect the validity of the tests though. ~~Not sure if it's just local to my machine~~ happens on github checks as well

**Comment by ypinhsuan** [comment]
I can't seem to find the cause of these errors. The `ngOnInit` of  `session-edit-page` is subscribing to an observable, so I tried to unsubscribe the observable in `ngOnDestroy` but it does not work. There are console errors for `question-edit-form` as well, not sure if they are related since `session-edit-page` uses `question-edit-form` component. I will continue to investigate further.

**Comment by ypinhsuan** [comment]
Hi, I need some help with the console errors. All the test works without error when run individually. The error is coming from these 5 tests: `should snap with new question added`, `should collapse all questions`, `should expand all questions when at least one question is collapsed`, `should create new question`. I tried searching online but could not find a solution that works for my case.

**Comment by ypinhsuan** [comment]
I managed to fix the console errors. It was because `ng-in-viewport` uses `IntersectionObserver`, so a mock observer was needed. It was fixed by adding a `disconnect` function to `jest-setup`.

**Review by halfwhole — APPROVED** [review]

**Comment by madanalogy** [comment]
@ypinhsuan seems like there's a merge conflict. Could you help resolve it please, thanks

**Comment by ypinhsuan** [comment]
I have resolved the merge conflicts 👍

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts [hunk 6]
```diff
@@ -1,33 +1,505 @@
...
+    spyOn(feedbackQuestionsService, 'getFeedbackQuestions').and.returnValue(of({ questions: [testFeedbackQuestion1] }));
+    spyOn(feedbackQuestionsService, 'createFeedbackQuestion').and.returnValue(of(testFeedbackQuestion1));
+
+    spyOn(ngbModal, 'open').and.returnValue(mockModalRef);
+
+    expect(component.copyCurrentSession()).resolves.toBe(undefined);
+    expect(ngbModal.open).toHaveBeenCalledWith(CopySessionModalComponent);
+    expect(mockModalRef.componentInstance.newFeedbackSessionName).toEqual(testFeedbackSession.feedbackSessionName);
+    expect(mockModalRef.componentInstance.courseCandidates[0]).toEqual(testCourse2);
+    expect(mockModalRef.componentInstance.sessionToCopyCourseId).toEqual(testCourse1.courseId);
```
Comment (rrtheonlyone) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:503]: What about testing this part?

```
modalRef.result.then((result: CopySessionModalResult) => {
        this.failedToCopySessions = {};
        const requestList: Observable<FeedbackSession>[] = this.createSessionCopyRequestsFromModal(
            result, this.courseId, this.feedbackSessionName);
        if (requestList.length === 1) {
          requestList[0].subscribe((createdSession: FeedbackSession) => {
            if (Object.keys(this.failedToCopySessions).length === 0) {
              this.navigationService.navigateWithSuccessMessage(this.router,
                  '/web/instructor/sessions/edit',
                  'The feedback session has been copied. Please modify settings/questions as necessary.',
                  { courseid: createdSession.courseId, fsname: createdSession.feedbackSessionName });
            } else {
              this.statusMessageService.showErrorToast(this.getCopyErrorMessage());
              this.sessionEditFormModel.isCopying = false;
            }
          }, (resp: ErrorMessageOutput) => { this.statusMessageService.showErrorToast(resp.error.message); });
        }
        if (requestList.length > 1) {
          forkJoin(requestList)
           .pipe(finalize(() => this.sessionEditFormModel.isCopying = false))
           .subscribe(() => {
             this.showCopyStatusMessage();
           });
        }
      }, (resp: ErrorMessageOutput) => { this.statusMessageService.showErrorToast(resp.error.message); })
      .catch(() => this.sessionEditFormModel.isCopying = false);
    });
```

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts [hunk 7]
```diff
@@ -1,33 +1,505 @@
...
+
+    component.loadFeedbackSession();
+
+    expect(component.courseName).toBe('Test Course 1');
+    expect(component.sessionEditFormModel.courseId).toBe('testId1');
+    expect(component.sessionEditFormModel.courseName).toBe('Test Course 1');
+    expect(component.sessionEditFormModel.feedbackSessionName).toBe('Test Session');
+  });
+
+  it('should load correct feedback session questions', () => {
```
Comment (rrtheonlyone) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:340]: What about testing the error cases for all these endpoints?

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Really nice work! Attached some comments below:

Also do look through and see if you have tested all the functionality e.g (save/duplicate/delete/move handlers)

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts (hunk 6 above)
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:701]: I have updated the test and tested this part by spying on navigation service

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts (hunk 7 above)
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts:374]: Added error tests for endpoints

**Review by rrtheonlyone — APPROVED** [review]
LGTM 💯  Really nice work!

**Pull request merged**